### PR TITLE
MAINT: Check for shadowing and mutable defaults

### PR DESCRIPTION
**Pull request by larsoner**
Enable rules to avoid shadowing builtins (should help with stuff like #12352) and mutable defaults. Found a real bug that could be out there in user code (but hopefully rare!) due to mutable defaults along with added test. Added exceptions/`noqa` for things like `format="nifti"`-style kwargs where it doesn't seem like a deprecation cost would be worth the gain of avoiding the shadowing.

Pushing with ci skip, will merge `main` into this PR once https://github.com/mne-tools/mne-python/pull/12377 lands. Can be reviewed in the meantime if someone wants, though.

### CODE REVIEW

File: doc/conf.py [hunk 1]
```diff
@@ -1382,6 +1382,7 @@ def reset_warnings(gallery_conf, fname):
         r"iteritems is deprecated.*Use \.items instead\.",
         "is_categorical_dtype is deprecated.*",
         "The default of observed=False.*",
+        "When grouping with a length-1 list-like.*",
```
Comment (larsoner) [inline doc/conf.py:1385]: Unrelated fix for https://app.circleci.com/pipelines/github/mne-tools/mne-python/22758/workflows/5b6b3106-2ca4-4ad2-8e68-0d63310cf410/jobs/63327

**Comment by drammock** [comment]
thank you for tackling this. The shadowed imports have been bugging me lately and I was hoping to get around to it this week. 🙏🏻

**Comment by larsoner** [comment]
> is it worth it to someday change the param names in the public API that are shadowing built-ins? e.g. id, format

I don't think so -- the code paths where we use those is small so low risk of doing something bad, we don't need for example `id` or `format` in them (and even if we did we could `from builtins import ...` under different names), and it will cause code churn for end users. So very small gain for some end user pain I think.
